design/ingressroute-design.md

```diff
@@ -0,0 +1,398 @@
+- Support non HTTP/HTTPS traffic ingress (ie, no UDP, no MongoDB, no TCP passthrough). We just do HTTP/HTTPS ingress.
+- Deprecate Contours support for v1beta1.Ingress.
+
+# Background
+
+The ingress object was added to kubernetes in 1.2 as a way of describing properties of a cluster wide reverse HTTP proxy.
+Since that time the ingress object has not progressed beyond the beta stage, and its stagnation inspired a cambrian explosion of annotations to express missing properties of http routing.
+
+# High-Level Design
+
+At a high level, this document proposes modeling ingress configuation as a graph of documents throughout a Kubernetes API server which when taken together form a directed acyclic graph (DAG) of the configuration for virtual hosts and their constituent routes.
+
+## Delegation
+
+The working model for delegation is DNS.
+As the owner of a DNS domain, for example `.com`, I _delegate_ to another nameserver the responsibility for handing the subdomain `heptio.com`.
+Any nameserver can hold a record for `heptio.com`, but without the linkage from the parent `.com` TLD, its information is unreachable and non authorative.
+
+Each _root_ of a DAG starts at a virtual host, which describes properties like the fully qualified name of the virtual host, any aliases (for example a www. prefix) of that vhost, TLS configuration, and possibly global access list details.
+The vertices of a graph do not contain virtual host information, they are only reachable from a root via a delegation.
+This permits the _owner_ of an ingress root to both delegate the authority to publish a service on a portion of the route space inside a virtual host, and to itself further delegate authority to publish and delegate.
+
+This also means that 
+
+In practice the linkage, or delegation, from root to vertex, is performed with a specific type of route action.
+You can think of it that rather than routing traffic to a service, it is routed to another ingressroute object for further processing.
+
+
+##
+
+# Detailed Design
+
+## IngressRoute CRD
+
+This is an example of a fully populated root ingressroute.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google
+  namespace: prod
+spec:
+  # virtualhost appears at most once. If it is present, the object is considered
+  # to be a "root".
+  virtualhost:
+    # the fully qualified domain name of the root of the ingress tree
+    # all leaves of the DAG rooted at this object relate to the fqdn (and its aliases)
+    fqdn: www.google.com
+    # a set of aliases for the domain, these may be alternative fqdns which are considered
+    # aliases of the primary fqdn
+    aliases:
+      - www.google.com.au
+      - google.com
+    # if present describes tls properties. The CNI names that will be matched on
+    # are described in fqdn and aliases, the tls.secretName secret must contain a
+    # matching certificate
+    tls:
```

Comment:
Initially I though that ideas like HTTP -> 301 -> HTTPS upgrade per a property of the route (that is how they are handled inside envoy) and that in addition to `spec.route.backend` and `spec.route.delegate` we could have some form of `spec.route.redirect` but that would require describing HTTP and HTTPS routes independently -- you want a route.redirect for port 80 and a route.backend for port 443. Again, this is how Envoy handles it internally.

However on reflection I think that it might be better to treat this specific requirement based on the behaviour the user wants, not the implementation. So it should be a field, `spec.virtualhost.auto-tls: true` which we interpret however we need to incorporate tools like cert-manager.

I have no hard evidence, but I believe the set of per-virtualhost options is limited, given the scope of the purview of ingress, that we can add this switches are required. Again this is informed by looking at the evolution of the Envoy http manager api which has magic flags like this.